mitiq/lre/multivariate_scaling/layerwise_folding.py

```diff
@@ -0,0 +1,199 @@
+# Copyright (C) Unitary Fund
+#
+# This source code is licensed under the GPL license (v3) found in the
+# LICENSE file in the root directory of this source tree.
+
+"""Functions for layerwise folding of input circuits to allow for multivariate
+extrapolation as defined in :cite:`Russo_2024_LRE`.
+"""
+
+import itertools
+from copy import deepcopy
+from typing import Any, Callable, List, Optional, Tuple
+
+import numpy as np
+from cirq import Circuit
+
+from mitiq import QPROGRAM
+from mitiq.utils import _append_measurements, _pop_measurements
+from mitiq.zne.scaling import fold_gates_at_random
+from mitiq.zne.scaling.folding import _check_foldable
+
+
+def _get_num_layers_without_measurements(input_circuit: Circuit) -> int:
+    """Checks if the circuit has non-terminal measurements and returns the
+    number of layers in the input circuit without the terminal measurements.
+
+        Args:
+            input_circuit: Circuit of interest
+
+        Returns:
+            num_layers: the number of layers in the input circuit without the
+                terminal measurements
+    """
+
+    _check_foldable(input_circuit)
```

Comment:
This can raise an UnfoldableCircuitError, it should be documented in the docstring of this caller function as well.

Reply:
I had included it previously. https://github.com/unitaryfund/mitiq/pull/2347#discussion_r1633629577

We should agree on style guidelines for this project and update [the docs style guidelines section](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines) on docstrings. 

Couple of things that confused me:

- Since I am relying on private functions, a user won't be able to read the source of the error in the API. Which is why I felt it's better to document the raised errors in the public function. 
- Referring to Vincent's comment on raising errors from a function in a different module, the API-doc page only mentions [`UnfoldableCircuitError`](https://mitiq.readthedocs.io/en/stable/apidoc.html#mitiq.zne.scaling.folding.UnfoldableCircuitError) once as a defined exception. As `_check_foldable` is a private function, [the error raised by the function](https://github.com/unitaryfund/mitiq/blob/2f9c00a69739605ff86785e88757233af17af748/mitiq/zne/scaling/folding.py#L66) does not show up publicly.  
- Most of the raised errors in the public-facing function are specified in a paragraph. 
![image](https://github.com/unitaryfund/mitiq/assets/66048318/9589a2f1-1476-4607-a726-0d8b9754cecc)
As we rely on google-style docstrings, there is no convention to allow lists in the description of the [different sections](https://google.github.io/styleguide/pyguide.html#doc-function-args) of a docstring. [Portions of our API-doc](https://mitiq.readthedocs.io/en/stable/apidoc.html#module-mitiq.cdr.cdr) have run-on sentences because someone used a list in the section description which appear as badly formatted sentences publicly.

| Intended      | in API-doc |
| ----------- | ----------- |
|![image](https://github.com/unitaryfund/mitiq/assets/66048318/d8133c18-eb0d-40a9-ae48-c1909fc5b7e2)| ![image](https://github.com/unitaryfund/mitiq/assets/66048318/b81f11c7-d1d3-4465-b533-aa73fc37e9e5)|